lib/Conversion/TorchOnnxToTorch/OnnxRnnExpander.cpp

```diff
@@ -0,0 +1,312 @@
+                                  ValueRange({Y_new, H_new}));
+  }
+  RnnLayerOutput output;
+  output.Y = loop.getResult(0);
+  output.Y_h = loop.getResult(1);
+  return output;
+}
+
+LogicalResult OnnxRnnExpander(OpBinder binder,
+                              ConversionPatternRewriter &rewriter) {
+  Location loc = binder.getLoc();
+  mlir::ImplicitLocOpBuilder b(loc, rewriter);
+
+  std::string direction;
+
+  ValueTensorType yTy, Y_hType;
+  if (binder.tensorResultTypeAtIndex(yTy, 0) ||
+      binder.tensorResultTypeAtIndex(Y_hType, 1)) {
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "At least one output must be present");
+  }
+  Value X;
+  if (binder.tensorOperandAtIndex(X, 0))
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "Missing required input tensor X");
+  Value W;
+  if (binder.tensorOperandAtIndex(W, 1))
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "Missing required input tensor W");
+  Value R;
+  if (binder.tensorOperandAtIndex(R, 2))
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "Missing required input tensor R");
+  int64_t hidden_size;
+  if (binder.s64IntegerAttr(hidden_size, "hidden_size"))
+    return rewriter.notifyMatchFailure(
+        binder.op, "Missing required attribute hidden_size");
+
+  auto xTy = cast<ValueTensorType>(X.getType());
+  auto wTy = cast<ValueTensorType>(W.getType());
+  Value B;
+  if (binder.tensorOperandAtIndex(B, 3)) {
+    B = b.create<AtenZerosOp>(W.getType(), W);
+  }
+
+  llvm::SmallVector<std::string> activationsList;
+
+  RnnActivations activations;
+  activations.f = "Tanh";
+
+  if (!binder.stringArrayAttr(activationsList, "activations") &&
+      activationsList.size() > 0) {
+    if (activationsList.size() == 1) {
+      activations.f = activationsList[0];
+    } else if (activationsList.size() == 2) {
+      return rewriter.notifyMatchFailure(
+          binder.op, "Bi-directional RNN is not yet supported, yet two "
+                     "activation function names are provided");
+    } else {
+      return rewriter.notifyMatchFailure(
+          binder.op, "Unsupported number of activation functions: " +
+                         std::to_string(activationsList.size()) +
+                         " are provided.");
+    }
+  }
+
+  if (!binder.customOpNameStringAttr(direction, "direction", "forward") &&
+      direction != "forward")
+    return rewriter.notifyMatchFailure(binder.op,
+                                       "Unsupported direction attribute value. "
+                                       "Only 'forward' is supported but '" +
+                                           direction + "' is provided.");
+  int64_t num_directions = (direction == "bidirectional") ? 2 : 1;
+
+  auto XShape = xTy.getSizes();
+  int64_t batch_size = XShape[1];
+  int64_t input_size = XShape[2];
```

Comment:
Does this conversion require static dims in order to work?

Reply:
I think so. I need static dims to construct the big result tensor storing the intermediate hidden states of each cell.